functions/database/database.js

```diff
@@ -0,0 +1,391 @@
+ */
+async function getLibraryId(libraryName) {
+  const librarySnapshot = await db.collection("libraries")
+      .where("libraryName", "==", libraryName)
+      .get();
+
+  if (librarySnapshot.empty) {
+    throw new Error(`
+      Library in release report does not exist in Firestore: ${libraryName}`,
+    );
+  }
+
+  return librarySnapshot.docs[0].id;
+}
+
+/**
+ * Deletes all existing change documents associated with a release.
+ *
+ * @param {admin.firestore.WriteBatch} batch The batch to add the
+ * delete operations to.
+ * @param {string} releaseId The ID of the associated release.
+ */
+async function batchDeleteReleaseChanges(batch, releaseId) {
+  const previousChangesSnapshot = await db.collection("changes")
+      .where("releaseID", "==", releaseId)
+      .get();
+  for (const doc of previousChangesSnapshot.docs) {
+    const docRef = db.collection("changes").doc(doc.id);
+    batch.delete(docRef);
+  }
+}
+
+/**
+ * Creates new change documents for a library and release.
+ *
+ * @param {admin.firestore.WriteBatch} batch The batch to add the
+ * delete operations to.
+ * @param {Object} changes The changes to add to Firestore.
+ * @param {string} libraryId The ID of the associated library.
+ * @param {string} releaseId The ID of the associated release.
+ */
+function batchSetReleaseChanges(batch, changes, libraryId, releaseId) {
+  changes.forEach((change) => {
+    const docRef = db.collection("changes").doc(change.commitId);
+    batch.set(docRef, {
+      commitTitle: parseCommitTitleFromMessage(change.message),
+      message: change.message,
+      author: "GitHub user", // change.author
+      pullRequestID: change.prId,
+      pullRequestLink: change.prLink,
+      commitID: change.commitId,
+      commitLink: change.commitLink,
+      libraryID: libraryId,
+      releaseID: releaseId,
+    });
+  });
+}
+
+/**
+ * Creates new change documents for each change in the release report, and
+ * deletes any existing changes associated with the release.
+ *
+ * @param {Object} releaseReport The release report containing changes by
+ * library name.
+ * @param {string} releaseId The ID of the associated release.
+ * @throws {Error} If a library in the release report does not exist in
+ * Firestore.
+ */
+async function updateChangesForRelease(releaseReport, releaseId) {
+  const batch = db.batch();
+
+  await batchDeleteReleaseChanges(batch, releaseId);
+
+  const changePromises = Object.entries(releaseReport.changesByLibraryName)
+      .map(async ([libraryName, changes]) => {
+        const libraryId = await getLibraryId(libraryName);
+        batchSetReleaseChanges(batch, changes, libraryId, releaseId);
+      });
+
+  await Promise.all(changePromises);
+
+  await batch.commit();
+}
+
+
+/**
+ * Deletes all existing check documents associated with a release.
+ *
+ * @param {admin.firestore.WriteBatch} batch The batch to add the
+ * delete operations to.
+ * @param {string} releaseId The ID of the associated release.
+ */
```

Comment:
Add a body here noting that this doesn't commit the change

```js
/**
 * Deletes all existing check documents associated with a release.
 *
 * Note that this will not actually commit the change, it merely adds
 * it to the provided batch.
 *
 * @param {admin.firestore.WriteBatch} batch The batch to add the
 * delete operations to.
 * @param {string} releaseId The ID of the associated release.
 */
```

Reply:
Resolved in https://github.com/firebase/firebase-release-dashboard/pull/8/commits/15c9d1d1cd01b3c812e1ca32042cbe083c0a0ce8